p2p/host/autonat/autonat.go

```diff
@@ -399,33 +399,30 @@ func (as *AmbientAutoNAT) getPeerToProbe() peer.ID {
 		return ""
 	}
 
-	candidates := make([]peer.ID, 0, len(peers))
+	// clean old probes
+	fixedNow := time.Now()
+	for k, v := range as.recentProbes {
+		if fixedNow.Sub(v) > as.throttlePeerPeriod {
+			delete(as.recentProbes, k)
+		}
+	}
 
-	for _, p := range peers {
+	n := len(peers)
+	for i, j := rand.Intn(n), 0; j < n; i, j = (i+1)%n, j+1 {
```

Comment:
I don't understand the point of starting at a random offset. First, the code is slightly confusing. But, more importantly, you probably want to actually shuffle the list, rather than starting at a random offset. Shuffling correctly is surprisingly [subtle](https://blog.codinghorror.com/the-danger-of-naivete/).

Reply:
We are only interested in the first element that satisfies our condition, so we I think it was fine to not shuffle the whole thing. 

Separately, theres `rand.Shuffle` if you do want to shuffle.

Reply:
We start at a random offset to not rely on the peerstore's ordering of peers.

Reply:
TIL about rand.Shuffle. We should make a PR that updates our usages of shuffling throughout the codebase.

I think we do want to shuffle here, although I don't think I could prove why.